src/lib/pubkey/ec_group/ec_inner_data.h

```diff
@@ -0,0 +1,230 @@
+/*
+* (C) 2024 Jack Lloyd
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#ifndef BOTAN_EC_INNER_DATA_H_
+#define BOTAN_EC_INNER_DATA_H_
+
+#include <botan/ec_group.h>
+
+#include <botan/asn1_obj.h>
+#include <botan/bigint.h>
+#include <botan/reducer.h>
+#include <botan/internal/point_mul.h>
+#include <botan/internal/stl_util.h>
+#include <span>
+
+namespace Botan {
+
+class EC_Scalar_Data final {
+   public:
+      EC_Scalar_Data(BigInt v) : m_v(std::move(v)) {}
+
+      std::unique_ptr<EC_Scalar_Data> clone() const { return std::make_unique<EC_Scalar_Data>(m_v); }
+
+      const BigInt& value() const { return m_v; }
+
+      void set_value(const BigInt& v) { m_v = v; }
+
+   private:
+      BigInt m_v;
+};
+
+class EC_Point_Data final {
+   public:
+      EC_Point_Data(EC_Point&& pt) : m_pt(std::move(pt)) {
+         m_pt.force_affine();
+         m_xy = m_pt.xy_bytes();
+      }
+
+      EC_Point_Data(const EC_Point& pt) : m_pt(pt) {
+         m_pt.force_affine();
+         m_xy = m_pt.xy_bytes();
+      }
+
+      const EC_Point& value() const { return m_pt; }
+
+      size_t field_element_bytes() const { return m_xy.size() / 2; }
+
+      void serialize_x_to(std::span<uint8_t> bytes) const {
+         const size_t fe_bytes = field_element_bytes();
+         BOTAN_ARG_CHECK(bytes.size() == fe_bytes, "Invalid output size");
+         copy_mem(bytes, std::span{m_xy}.first(fe_bytes));
+      }
+
+      void serialize_y_to(std::span<uint8_t> bytes) const {
+         const size_t fe_bytes = field_element_bytes();
+         BOTAN_ARG_CHECK(bytes.size() == fe_bytes, "Invalid output size");
+         copy_mem(bytes, std::span{m_xy}.last(fe_bytes));
+      }
+
+      void serialize_xy_to(std::span<uint8_t> bytes) const {
+         const size_t fe_bytes = field_element_bytes();
+         BOTAN_ARG_CHECK(bytes.size() == 2 * fe_bytes, "Invalid output size");
+         copy_mem(bytes, m_xy);
+      }
+
+      void serialize_compressed_to(std::span<uint8_t> bytes) const {
+         const size_t fe_bytes = field_element_bytes();
+         BOTAN_ARG_CHECK(bytes.size() == 1 + fe_bytes, "Invalid output size");
+         const bool y_is_odd = (m_xy[m_xy.size() - 1] & 0x01) == 0x01;
+
+         BufferStuffer stuffer(bytes);
+         stuffer.append(y_is_odd ? 0x03 : 0x02);
+         stuffer.append(std::span{m_xy}.last(fe_bytes));
```

Comment:
See also: https://github.com/randombit/botan/pull/4042#discussion_r1635928550

Reply:
LOL WTF that definitely should have caused test failures, need to look at this

Reply:
Ah OK the reason this was missed is that currently this serialization logic in unused, because keys/etc are still using `EC_Point`.

We're in for a somewhat awkward situation in the short term as we'll have 3 distinct elliptic curve point types - original flavor `EC_Point`, this `EC_AffinePoint`, plus the (internal only) pcurves `AffinePoint`, with a lot of redundant/duplicated logic.

For 3.6.0 onward the situation improves

* The various pubkey algorithms switch to storing their points as `EC_AffinePoint`. At that point `EC_Point` becomes unused except for testing; unfortunately we can't remove it completely until Botan 4. Maybe we'll be able to segregate it into a deprecated submodule of `ec_group` though.

* `EC_AffinePoint` gains a bridge to pcurve. This affords removing some redundancies between pcurves and what we actually need in practice. For example pcurves has `scalar_from_bits_with_trunc` but this is just modular reduction plus some shifts. It'll likely be simpler, once the bridge exists, to remove that and do the shifting in some higher point in the call stack so the logic is shared between the pcurves and BigInt based approaches.